core/src/display_object/movie_clip.rs

```diff
@@ -1000,6 +1374,204 @@ impl<'gc> MovieClip<'gc> {
+                );
+
+                for key in init_object.get_keys(&mut activation) {
+                    if let Ok(value) = init_object.get(&key, &mut activation) {
+                        let _ = object.set(&key, value, &mut activation);
+                    }
+                }
+            }
+            let mut mc = self.0.write(context.gc_context);
+            mc.object = Some(object.into());
+
+            let mut events = Vec::new();
+
+            for clip_action in mc
+                .clip_actions()
+                .iter()
+                .filter(|action| action.event == ClipEvent::Construct)
+            {
+                events.push(clip_action.action_data.clone());
+            }
+
+            context.action_queue.queue_actions(
+                display_object,
+                ActionType::Construct {
+                    constructor: mc.avm_constructor.map(|a| a.as_avm1_object().unwrap()),
+                    events,
+                },
+                false,
+            );
+        }
+
+        if run_frame {
+            self.run_frame(context);
+        }
+
+        // If this text field has a variable set, initialize text field binding.
+        Avm1::run_with_stack_frame_for_display_object(
+            self.into(),
+            context.swf.version(),
+            context,
+            |activation| {
+                self.bind_text_field_variables(activation);
+            },
+        );
+    }
+
+    fn construct_as_avm2_object(
+        self,
+        context: &mut UpdateContext<'_, 'gc, '_>,
+        display_object: DisplayObject<'gc>,
+    ) {
+        let constructor = self.0.read().avm_constructor;
+        if let Some(AvmObject::Avm2(mut constr)) = constructor {
+            let mut constr_thing = || {
+                let mut activation = Avm2Activation::from_nothing(context.reborrow());
+                let proto = constr
+                    .get_property(
+                        constr,
+                        &Avm2QName::new(Avm2Namespace::public_namespace(), "prototype"),
+                        &mut activation,
+                    )?
+                    .coerce_to_object(&mut activation)?;
+                let object = Avm2StageObject::for_display_object(
+                    activation.context.gc_context,
+                    display_object,
+                    proto,
+                )
+                .into();
+
+                constr.call(Some(object), &[], &mut activation, Some(proto))?;
+
+                Ok(object)
+            };
+            let result: Result<Avm2Object<'gc>, Avm2Error> = constr_thing();
+
+            if let Ok(object) = result {
+                self.0.write(context.gc_context).object = Some(object.into());
+            } else if let Err(e) = result {
+                log::error!("Got {} when constructing AVM2 side of display object", e);
+            }
+        } else {
+            log::error!("Attempted to construct AVM2 movieclip without constructor!");
```

Comment:
This error gets emitted by any standard MC on the timeline; is that right? Should it default to the MovieClip constructor?

Reply:
I believe you're right, it should just be using `MovieClip` if the SWF doesn't say otherwise.